lib/ecto/repo/schema.ex

```diff
@@ -31,7 +31,12 @@ defmodule Ecto.Repo.Schema do
     end
   end
 
-  defp do_insert_all(name, schema, prefix, source, rows, opts) when is_list(rows) do
+  defp do_insert_all(name, schema, prefix, source, rows, opts) do
+    case rows do
+      list when is_list(list) -> :ok
+      %Ecto.Query{} -> :ok
+      _ -> raise ArgumentError, message: "insert_all called with invalid rows_or_query: #{inspect rows}"
+    end
```

Comment:
Btw, something else you can do, and perhaps cleaner, is to check for this on `extract_header_and_fields`.

Reply:
yes, excellent idea